pkg/controller/rollout/trafficrouting/gateway/gateway.go

```diff
@@ -121,13 +111,67 @@ func (r *gatewayController) Finalise(ctx context.Context) error {
 		return r.Client.Update(context.TODO(), routeClone)
 	}); err != nil {
 		klog.Errorf("update rollout(%s/%s) httpRoute(%s) failed: %s", r.conf.RolloutNs, r.conf.RolloutName, httpRoute.Name, err.Error())
-		return err
+		return false, err
 	}
 	klog.Infof("rollout(%s/%s) TrafficRouting Finalise success", r.conf.RolloutNs, r.conf.RolloutName)
-	return nil
+	return false, nil
+}
+
+func (r *gatewayController) buildDesiredHTTPRoute(rules []gatewayv1alpha2.HTTPRouteRule, weight *int32, matches []rolloutv1alpha1.HttpRouteMatch) []gatewayv1alpha2.HTTPRouteRule {
+	var desired []gatewayv1alpha2.HTTPRouteRule
+	// Only when finalize method parameter weight=-1, then we need to remove the canary route policy and restore to the original configuration
+	if weight != nil && *weight == -1 {
```

Comment:
can we replace hard coded special value -1 with nil ?

Reply:
Cannot, because step.weight=nil is valid parameter. weight=-1 specifically refers to the need to delete the canary configuration.